host_core/lib/host_core/application.ex

```diff
@@ -26,4 +32,67 @@ defmodule HostCore.Application do
     Logger.info("Starting Host Core")
     Supervisor.start_link(children, opts)
   end
+
+  defp retrieve_nats_env() do
+    rpc_host =
+      case System.get_env("WASMCLOUD_RPC_HOST") do
+        nil -> '0.0.0.0'
+        host -> host
+      end
+
+    rpc_port =
+      case System.get_env("WASMCLOUD_RPC_PORT") do
+        nil -> 4222
+        port -> String.to_integer(port)
+      end
+
+    ctl_host =
+      case System.get_env("WASMCLOUD_CTL_HOST") do
+        nil -> '0.0.0.0'
+        host -> host
+      end
+
+    ctl_port =
+      case System.get_env("WASMCLOUD_CTL_PORT") do
+        nil -> 4222
+        port -> String.to_integer(port)
+      end
+
+    {rpc_host, rpc_port, ctl_host, ctl_port}
+  end
+
+  defp start_gnat({rpc_host, rpc_port, ctl_host, ctl_port}) do
+    :ok = lattice_connect(rpc_host, rpc_port)
+    :ok = control_connect(ctl_host, ctl_port)
+  end
+
+  defp lattice_connect(rpc_host, rpc_port) do
+    case Gnat.start_link(%{host: rpc_host, port: rpc_port}, name: :lattice_nats) do
+      {:ok, _gnat} ->
+        :ok
+
+      {:error, :econnrefused} ->
+        Logger.warn(
+          "Failed to connect to lattice at #{rpc_host}:#{rpc_port}, retrying in 1 second..."
+        )
+
+        Process.sleep(1000)
```

Comment:
Rather than sleep, you should `send_after` which would then trigger a `handle_info` that responds to the message you sent to yourself. This is far more efficient (and concurrent-friendly) than sleeping the scheduler thread. Check out [send_after](https://hexdocs.pm/elixir/1.12/Process.html#send_after/3)

Reply:
Would this require refactoring the nats connection logic into a GenServer module? From some reading it looks like `handle_info` is a GenServer thing, not an Application thing (but I may be mistaken)

Reply:
yeah you generally want to keep all logic out of the `Application` entirely, and _just_ specify a list of children. I think an `Application` may actually be a `GenServer`, but it feels cleaner to have the connection be managed by a GenServer. We could probably put the `Gnat.connect` inside the `Host` GenServer or one of the other supervisors that automatically gets started.

Reply:
This will become a bit more tricky once we implement the lattice control interface, which will need to create subscriptions... so we'll have to ensure that those subscriptions aren't created until after the Gnat connection is managed. This leads me to think maybe we need a supervisor that tries to create a nats connection and then when it finally does, it spawns the `GnatServer` used for the request/reply features of the lattice control interface.

Reply:
I was doing some thinking about this, the original reason why I threw that logic in the Application was because there were a few children that make subscriptions, so I figured that there was no point in starting the children before Gnat was initialized. I am thinking something similar to your suggestion in your second comment:
> a supervisor that tries to create a nats connection and then when it finally does, it spawns the GnatServer used for the request/reply features of the lattice control interface
For example, a GnatSupervisor (name WIP) that handles creating the connection for the lattice / control interface, and then subscriptions are formed by sending a message to that supervisor with the PID of the interested GenServer.

For example, in `HostCore.LinkdefsManager`:
```
    {:ok, _sub} = Gnat.sub(:lattice_nats, self(), topic)
```
Would then change to something like
```
HostCore.GnatSupervisor.sub(:lattice_nats, self(), topic)
```

`HostCore.GnatSupervisor.sub/3` would accept a pid/atom of the connection, a pid of the interested genserver, and the topic. 

My biggest concern is ensuring that Gnat calls don't occur while we're waiting to establish the connection, I'm not sure if there's an option to start children in-order, let me know if I'm just missing that entirely